flows/vision-zero/cris_import/lib/sql.py

```diff
@@ -0,0 +1,460 @@
+    where 
+    {public_key_sql}
+    {linkage_sql}
+    """
+    return sql
+
+
+def form_insert_statement(
+    output_map, table, input_column_names, import_key_sql, DB_IMPORT_SCHEMA
+):
+    sql = f"insert into public.{output_map[table]} "
+    sql += "(" + ", ".join(input_column_names) + ") "
+    sql += "(select "
+    sql += ", ".join(input_column_names)
+    sql += f" from {DB_IMPORT_SCHEMA}.{table}"
+    sql += f" where {import_key_sql})"
+    return sql
+
+
+def is_change_existing(pg, record_type, record_id):
+    #print(f"\b 🛎 is existing? {record_type}, {record_id}")
+
+    sql = f"""
+    select count(*) as exists
+    from atd_txdot_changes_view
+    where record_id = {record_id}
+    """
+
+    cursor = pg.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+    cursor.execute(sql)
+    change_request_exists = cursor.fetchone()
+    if change_request_exists["exists"] > 0:
+        return True
+    else:
+        return False
+
+def has_existing_temporary_record(pg, case_id):
+    sql = f"""
+    select count(*) as exists
+    from atd_txdot_crashes
+    where crash_id < 10000
+    and case_id = '{case_id}'
+    """
+    cursor = pg.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+    cursor.execute(sql)
+    change_request_exists = cursor.fetchone()
+    if change_request_exists["exists"] > 0:
+        return True
+    else:
+        return False
+
+def remove_existing_temporary_record(pg, case_id):
+    sql = f"""
+    delete from atd_txdot_crashes
+    where crash_id < 10000
+    and case_id = '{case_id}'
+    """
+    cursor = pg.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+    cursor.execute(sql)
+    return True
+
+def try_statement(pg, output_map, table, public_key_sql, sql, dry_run):
+    if dry_run:
+        print("Dry run; skipping")
+        return
+    try:
+        cursor = pg.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+        cursor.execute(sql)
+        pg.commit()
+    except Exception as error:
+        print(
+            f"There is likely an issue with existing data. Try looking for results in {output_map[table]} with the following WHERE clause:\n'{public_key_sql}'"
+        )
+        print(f"Error executing:\n\n{sql}\n")
+        print("\a")  # 🛎
+
+
+def get_input_column_names(pg, DB_IMPORT_SCHEMA, table, target_columns):
+    sql = f"""
+    SELECT
+        column_name,
+        data_type,
+        character_maximum_length AS max_length,
+        character_octet_length AS octet_length
+    FROM
+        information_schema.columns
+    WHERE true
+        AND table_schema = '{DB_IMPORT_SCHEMA}'
+        AND table_name = '{table}'
+    """
+
+    cursor = pg.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+    cursor.execute(sql)
+    input_table_column_types = cursor.fetchall()
+
+    # TODO figure out how to .map() this
+    input_column_names = []
+    for result in input_table_column_types:
+        input_column_names.append(result["column_name"])
```

Comment:
[List comprehensions](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions) are fun for this but this might be more readable as is 👍